fc/parsing/actions.py

```diff
@@ -1200,204 +1266,301 @@ def _number_generator(self, value, units):
-                    units = self.units.get_unit(output.units)
-                    self.time_variable = self.model.convert_variable(variable, units, DataDirectionFlow.INPUT)
-                    break
+        # Add outputs
+        #   output <prefix:term> [units <uname>]
+        for ref in self.outputs:
+            if ref.prefixed_name == 'oxmeta:state_variable':    # Skip 'magic' oxmeta:state_variable term
+                continue
+            pvar = get(ref.prefixed_name)
+            pvar.output = True
+            pvar.merge_units(ref.units)
+
+        # Add optional declarations
+        #   optional <prefix:term> [default <simple_expr>]
+        for ref in self.optional_decls:
+            pvar = get(ref.prefixed_name)
+            pvar.optional = True
+            pvar.merge_default_expr(ref.default_expr)
+
+        # Add local variables
+        #   var <name> units <uname> [= <initial_value>]
+        for ref in self.local_var_declarations:
+            # Local variable names must be unique, and can't even be re-used in imported/nested protocols
+            if ref.name in self.protocol_variables:
+                raise ProtocolError(f'Variable "{ref.name}" was defined by more than one var statement.')
 
-    def _create_variable(self, var):
-        """Creates and returns an annotated variable based on the given :class:`VariableReference`."""
+            # Create and store variable
+            pvar = self.protocol_variables[ref.name] = ProtocolVariable(ref.name)
+            pvar.local = True
+            pvar.units = ref.units
+            pvar.initial_value = ref.initial_value
 
-        name = self.model.get_unique_name('protocol__' + var.local_name)
-        units = self.units.get_unit(var.units)
-        variable = self.model.add_variable(name, units)
-        self.model.add_cmeta_id(variable)
-        self.model.rdf.add((variable.rdf_identity, PRED_IS, var.rdf_term))
-        return variable
+        # Store equations from define and clamp statements
+        for eq in self.equations:
+            name = eq.var.name()
+            try:
+                pvar = self.protocol_variables[name]
+            except KeyError:
+                # Variable not found: Still OK, as long as it refers to an existing model variable
+                try:
+                    self._variable_generator(name)
+                except MissingVariableError:
+                    raise ProtocolError(f'Define or clamp statement found for unknown variable: {name}.')
 
-    def _add_input_variables(self):
-        """Ensure requested input variables exist, unless they are optional.
+                # Create new protocol variable to store info
+                pvar = self.protocol_variables[name] = ProtocolVariable(name)
 
-        If the variable exists its units are checked and converted if needed.
+            # Any variable that is redefined by an equation is treated as an input
```

Comment:
My reasoning was that if you redefine a variable, (and you suggested on slack that this could be done for inputs, outputs, or anything in the protocol), then it needs to be units-converted using the `input` rather than `output` rule.
Also, we're changing its value, which to me makes it an input.
(Similarly I'd argue time is an input)

Reply:
No, input here is a technical term that means specifically that the rest of the protocol (outside the model interface) is allowed to change the value of the variable at arbitrary time points. Which is why unit conversion for inputs needs to be handled differently: you need to ensure that when the protocol does this, the rest of the model sees the unit-converted value. Time *is* implicitly set from the protocol (via the solver) so does get converted using the input rule.

But arbitrary `define` clauses don't necessarily need or want this special processing. And except for ODEs or constants, it's actually inconsistent to have a `define` with an input on the LHS, because that results in an over-defined variable. States can change instantaneously, variables defined always to equal the result of a calculation can't. A define clause isn't changing the value of a variable at a particular time point, it's changing the always-true mathematical relationships between variables in the model.

Reply:
OK, then I think I can just remove this line!